File: ruby/lesson 6/in_order_iterative.rb
# https://leetcode.com/problems/binary-tree-inorder-traversal/

def inorder_traversal(root)
    list = []
    stack = []
    current = root
    while !current.nil? || !stack.empty? do
        if !current.nil?
            stack.push(current)
            current = current.left
        else
            current = stack.pop()
            list.push(current.val)
            current = current.right
        end
    end
    list
end

File: java/lesson 6/InOrderIterative.java
// https://leetcode.com/problems/binary-tree-inorder-traversal/

class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();        
        Stack<TreeNode> stack = new Stack<>();
        TreeNode current = root;
        while (current != null || !stack.empty()) {
            if (current != null) {
                stack.push(current);
                current = current.left;
            } else {
                current = stack.pop();
                list.add(current.val);
                current = current.right;
            }
        }
        return list;
    }
}
